File: python/0031_NextPermutation.py
class Solution(object):
    def nextPermutation(self, nums):
        """
        :type nums: List[int]
        :rtype: None Do not return anything, modify nums in-place instead.
        """
        i=j=len(nums)-1
        while i>0 and nums[i-1]>=nums[i]:
            i-=1
        if i==0:
            nums[:]=nums[::-1]
        else:
            while nums[j] <= nums[i-1]:
                j -= 1
            nums[i-1], nums[j] = nums[j], nums[i-1]
            nums[i:]= nums[len(nums)-1:i-1:-1]

File: cpp/0031_NextPermutation.cpp

class Solution {
public:
    void nextPermutation(vector<int>& nums) {
        int n=nums.size();
        for(int i=n-1;i>0;i--)
        {
            if(nums[i-1]<nums[i])
            {

                int index=findReplace(nums, i,n-1,nums[i-1]);
                int tmp=nums[index];
                nums[index]=nums[i-1];
                nums[i-1]=tmp;
                sort(nums,i,n-1);
                return;

            }

        }
        sort(nums,0,n-1);

    }
    void sort(vector<int> & nums,int left,int right)
    {
        while(right>left)
        {
            int tmp=nums[right];
            nums[right]=nums[left];
            nums[left]=tmp;
            right--;
            left++;
        }
    }

    int findReplace(vector<int> & nums,int left,int right,int num)
    {
        while(right>left)
        {
            int mid=left+(right-left)/2;
            if(nums[mid]>num)
            {
                left=mid+1;                
            }
            else
            {
                right=mid-1;
            }
        }
        if(nums[right]>num)
            return right;
        else
            return right-1;

    }

};
